crates/ide/src/syntax_highlighting/asm.rs

```diff
@@ -0,0 +1,667 @@
+//! Syntax highlighting for asm macro strings.
+use ide_db::SymbolKind;
+use syntax::{
+    ast::{self, FormatSpecifier, HasFormatSpecifier},
+    AstNode, AstToken, TextRange,
+};
+
+use crate::{syntax_highlighting::highlights::Highlights, HlRange, HlTag};
+
+enum AsmPart {
+    Comment,
+    Directive,
+    FormatClose,
+    FormatIdentifier,
+    FormatOpen,
+    Instruction,
+    Label,
+    NumericLiteral,
+    Register,
+}
+
+pub(super) fn highlight_asm_string(stack: &mut Highlights, string: &ast::String, range: TextRange) {
+    if is_asm_string(string).is_none() {
+        return;
+    }
+
+    log::info!("Highlighting asm string");
+    let _timer = stdx::timeit("highlighting asm string");
+    lex_asm_string(string, |piece_range, kind| {
+        stack.add(HlRange {
+            range: piece_range + range.start(),
+            highlight: highlight_asm_part(kind).into(),
+            binding_hash: None,
+        });
+    });
+}
+
+fn is_asm_string(string: &ast::String) -> Option<()> {
+    let parent = string.syntax().parent()?;
+
+    let name = parent.parent().and_then(ast::MacroCall::cast)?.path()?.segment()?.name_ref()?;
+    if !matches!(name.text().as_str(), "asm" | "global_asm") {
+        return None;
+    }
+
+    let first_literal = parent
+        .children_with_tokens()
+        .find_map(|it| it.as_token().cloned().and_then(ast::String::cast))?;
+
+    if &first_literal != string {
+        return None;
+    }
+
+    Some(())
+}
+
+fn highlight_asm_part(kind: AsmPart) -> HlTag {
+    match kind {
+        AsmPart::Comment => HlTag::Comment,
+        AsmPart::Directive => HlTag::Attribute,
+        AsmPart::FormatIdentifier => HlTag::Symbol(SymbolKind::Local),
+        AsmPart::FormatOpen | AsmPart::FormatClose => HlTag::FormatSpecifier,
+        AsmPart::Instruction => HlTag::Symbol(SymbolKind::Function),
+        AsmPart::Label => HlTag::Symbol(SymbolKind::LifetimeParam),
+        AsmPart::NumericLiteral => HlTag::NumericLiteral,
+        AsmPart::Register => HlTag::Symbol(SymbolKind::Static),
+    }
+}
+
+fn lex_asm_string<F>(string: &ast::String, mut callback: F)
```

Comment:
Would you like to publish & maintain a crate for syntax analysis of rust-flavored asm? ^^ It seems like this is a generally re-usable bit of functionality.

Reply:
Oh wait, right, the syntax is arch-specific... Hm, I guess, you know more about this than I do, but do consider if we could salvage some re-usability here?

Reply:
I wouldn't be opposed to extracting it out into its own crate(s) :) I think it could probably be its own crate since the formats aren't usually particularly complicated so each arch would probably just be a module.

Reply:
oh, I meant to ask: what did you have in mind with "syntax analysis of rust-flavored asm"? I'm assuming that includes tracking labels and such between different `asm!` blocks to try and give more complete information to the user? is there anything else? I suppose it may also be possible to give some information about the "type" of labels and such as well :thinking:

Reply:
I don’t know what is this exactly, but I imagine you need to pars asm to do things like:

* completion of instructions
* instruction explanations in hover
* Highlighting targets of jumps
* Highlighting syntax errors in asm blocks